contractsV5/KyberNetwork.sol

```diff
@@ -523,28 +524,107 @@ contract KyberNetwork is Withdrawable, Utils, IKyberNetwork, ReentrancyGuard {
+        tradeData.totalFeeWei = tradeData.tradeWeiAmount * takerFeeBps / BPS * tradeData.feePayingPercentage / 100;
+        uint tradeWeiAmountAfterFee = tradeData.tradeWeiAmount - tradeData.feeInWei;        
 
+        //reset amountSoFarNoFee
+        //no need to reset percentageSoFar, as the only case to handle is after searching for best ETH -> token reserve
+        amountSoFarNoFee = 0;
 
         // Eth to token
         ///////////////
         // if hinted reserves, find rates and save.
+        if (tradeData.ethToToken.addresses.length > 0) {
+            for (i = 0; i < tradeData.ethToToken.addresses.length; i++) {
+                IKyberReserve reserve = tradeData.ethToToken.addresses[i];
+                //calculate split amount, with and without fee
+                splitAmountAfterFee = (i == tradeData.ethToToken.splitValuesPercent.length-1) ? (tradeWeiAmountAfterFee - amountSoFarAfterFee) : tradeData.ethToToken.splitValuesPercent[i] * tradeWeiAmountAfterFee /  100;
+                splitAmountNoFee = (i == tradeData.ethToToken.splitValuesPercent.length-1) ? (tradeWeiAmount - amountSoFarNoFee) : tradeData.ethToToken.splitValuesPercent[i] * tradeWeiAmount /  100;
+                //calculate split and corresponding trade amounts
+                amountSoFarAfterFee += splitAmountAfterFee;
+                amountSoFarNoFee += splitAmountNoFee;
+                //to save gas, we make just 1 conversion rate call with splitAmountAfterFee
+                rate = reserve.getConversionRate(src, dest, splitAmountAfterFee, block.number);
+
+                //using same rate, calculate both destAmounts (with and without fees)
+                tradeData.destAmount += calcDestAmountWithDecimals(ETH_DECIMALS, tradeData.ethToToken.decimals, splitAmountAfterFee, rate);
+                destAmountNoFee += calcDestAmountWithDecimals(ETH_DECIMALS, tradeData.ethToToken.decimals, splitAmountNoFee, rate);
+
+                //save rate data
+                tradeData.ethToToken.rates[i] = rate;
+            }
+        } else {
+            // else, search best reserve and its rate
+            // issue: To save gas, I have to search best rate with token -> ETH fee deduction
+            (reserve, rate, isFeePaying) = searchBestRate(ETH_TOKEN_ADDRESS, dest, tradeWeiAmountAfterFee, tradeData.usePermissionless);
+            // deduct fee from tradeWeiAmount if necessary
+            if (isFeePaying) {
+                tradeWeiAmountAfterFee -= tradeWeiAmountAfterFee * takerFeeBps / BPS;
+                tradeData.feePayingPercentage += 100; //max percentage amount for ETH -> token
+            }
 
-        // else find best rate
-        // calculate accumulated amounts
-
-
-        // dest amount sbutract total fee wei
-
-        // calc final rate
+            //calculate destAmounts (with and without fees)
+            tradeData.destAmount = calcDestAmountWithDecimals(ETH_DECIMALS, tradeData.ethToToken.decimals, tradeWeiAmountAfterFee, rate);
```

Comment:
can we calculate this with
(tradeWeiAmount - weiFeeAmount)
and avoid local parameter?

Reply:
I assume you're referring to `tradeWeiAmountAfterFee`. We can, but `tradeWeiAmountAfterFee` is used quite a number of times. 6, to be exact.